File: CleverFerretV2/core/common/src/main/java/com/cleverferret/v2/core/common/result/IntegrationResult.java
package com.cleverferret.v2.core.common.result;

import com.cleverferret.v2.core.common.error.AppError;
import com.cleverferret.v2.core.common.error.UserFacingState;

public sealed interface IntegrationResult<T> permits IntegrationResult.Success, IntegrationResult.Failure {

    record Success<T>(T value) implements IntegrationResult<T> {}

    record Failure<T>(AppError error, UserFacingState state) implements IntegrationResult<T> {}

    static <T> Success<T> success(T value) {
        return new Success<>(value);
    }

    static <T> Failure<T> failure(AppError error, UserFacingState state) {
        return new Failure<>(error, state);
    }
}

File: CleverFerretV2/core/common/src/main/java/com/cleverferret/v2/core/common/result/IntegrationResult.kt
package com.cleverferret.v2.core.common.result

import com.cleverferret.v2.core.common.error.AppError
import com.cleverferret.v2.core.common.error.UserFacingState

sealed interface IntegrationResult<out T> {
    data class Success<T>(val value: T) : IntegrationResult<T>
    data class Failure(val error: AppError, val state: UserFacingState) : IntegrationResult<Nothing>

    companion object {
        fun <T> success(value: T): IntegrationResult<T> = Success(value)
        fun failure(error: AppError, state: UserFacingState): IntegrationResult<Nothing> = Failure(error, state)
    }
}
